api/rest/bundle_handler.go

```diff
@@ -0,0 +1,350 @@
+		return
+	}
+
+	dataFile, err := os.Create(filepath.Join(h.workDir, id, dataFileName))
+	if err != nil {
+		writeJSONError(w, http.StatusInsufficientStorage, fmt.Errorf("could not create data file %s: %s", id, err))
+	}
+
+	//TODO(janisz): use context cancel function to cancel bundle creation https://jira.mesosphere.com/browse/DCOS_OSS-5222
+	ctx, _ := context.WithTimeout(context.Background(), h.bundleCreationTimeout) //nolint:govet
+	done := make(chan []string)
+
+	go collectAll(ctx, done, dataFile, h.collectors)
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			break
+		case bundle.Errors = <-done:
+			bundle.Status = Done
+			bundle.Stopped = h.clock.Now()
+			if e := ioutil.WriteFile(stateFilePath, jsonMarshal(bundle), filePerm); e != nil {
+				logrus.WithError(e).Errorf("Could not update state file %s", id)
+			}
+		}
+	}()
+
+	write(w, bundleStatus)
+}
+
+func collectAll(ctx context.Context, done chan<- []string, dataFile io.WriteCloser, collectors []collector.Collector) {
+	zipWriter := zip.NewWriter(dataFile)
+	var errors []string
+	// summaryReport is a log of a diagnostics job
+	summaryReport := new(bytes.Buffer)
+
+	for _, c := range collectors {
+		if ctx.Err() != nil {
+			errors = append(errors, ctx.Err().Error())
+			break
+		}
+		summaryReport.WriteString(fmt.Sprintf("[START GET %s]\n", c.Name()))
+		err := collect(ctx, c, zipWriter)
+		summaryReport.WriteString(fmt.Sprintf("[STOP GET %s]\n", c.Name()))
+		if err != nil && !c.Optional() {
+			errors = append(errors, err.Error())
+		}
+	}
+
+	summaryReportFile, err := zipWriter.Create(summaryReportFileName)
+	if err != nil {
+		errors = append(errors, err.Error())
+	} else {
+		if _, err := io.Copy(summaryReportFile, summaryReport); err != nil {
+			errors = append(errors, err.Error())
+		}
+	}
+
+	if len(errors) != 0 {
+		summaryErrorReportFile, err := zipWriter.Create(summaryErrorsReportFileName)
+		if err != nil {
+			errors = append(errors, err.Error())
+		} else {
+			if _, err := summaryErrorReportFile.Write([]byte(strings.Join(errors, "\n"))); err != nil {
+				errors = append(errors, err.Error())
+			}
+		}
+	}
+
+	if err := zipWriter.Close(); err != nil {
+		errors = append(errors, err.Error())
+	}
+	if err := dataFile.Close(); err != nil {
+		errors = append(errors, err.Error())
+	}
+
+	done <- errors
+}
+
+func collect(ctx context.Context, c collector.Collector, zipWriter *zip.Writer) error {
+	rc, err := c.Collect(ctx)
+	if err != nil {
+		return fmt.Errorf("could not collect %s: %s", c.Name(), err)
+	}
+	defer rc.Close()
+
+	zipFile, err := zipWriter.Create(c.Name())
+	if err != nil {
+		return fmt.Errorf("could not create a %s in the zip: %s", c.Name(), err)
+	}
+	if _, err := io.Copy(zipFile, rc); err != nil {
+		return fmt.Errorf("could not copy data to zip: %s", err)
+	}
+
+	return nil
+}
+
+func (h BundleHandler) Get(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+
+	bundle, err := h.getBundleState(id)
```

Comment:
Same as comment before. I think a separate `BundleExists(id)` method would make this much clearer.